contracts/state/StateV2.sol

```diff
@@ -0,0 +1,464 @@
+            }
+        }
+        return (0, 0, 0);
+    }
+
+    /**
+     * @dev binary search by timestamp
+     * @param id identity
+     * @param timestamp timestamp
+     * return parameters are (by order): block number, block timestamp, state
+     */
+    function getStateDataByTime(uint256 id, uint64 timestamp)
+        public
+        view
+        returns (
+            uint64,
+            uint64,
+            uint256
+        )
+    {
+        require(timestamp < block.timestamp, "errNoFutureAllowed");
+        // Case that there is no state committed
+        if (identities[id].length == 0) {
+            return (0, 0, 0);
+        }
+        // Case that there timestamp searched is beyond last timestamp committed
+        uint64 lastTimestamp = identities[id][identities[id].length - 1]
+            .BlockTimestamp;
+        if (timestamp > lastTimestamp) {
+            return (
+                identities[id][identities[id].length - 1].BlockN,
+                identities[id][identities[id].length - 1].BlockTimestamp,
+                identities[id][identities[id].length - 1].State
+            );
+        }
+        // Binary search
+        uint256 min = 0;
+        uint256 max = identities[id].length - 1;
+        while (min <= max) {
+            uint256 mid = (max + min) / 2;
+            if (identities[id][mid].BlockTimestamp == timestamp) {
+                return (
+                    identities[id][mid].BlockN,
+                    identities[id][mid].BlockTimestamp,
+                    identities[id][mid].State
+                );
+            } else if (
+                (timestamp > identities[id][mid].BlockTimestamp) &&
+                (timestamp < identities[id][mid + 1].BlockTimestamp)
+            ) {
+                return (
+                    identities[id][mid].BlockN,
+                    identities[id][mid].BlockTimestamp,
+                    identities[id][mid].State
+                );
+            } else if (timestamp > identities[id][mid].BlockTimestamp) {
+                min = mid + 1;
+            } else {
+                max = mid - 1;
+            }
+        }
+        return (0, 0, 0);
+    }
+
+    /**
+     * @dev Retrieve identity last committed information
+     * @param id identity
+     * @return last state for a given identity
+     * return parameters are (by order): block number, block timestamp, state
+     */
+    function getStateDataById(uint256 id)
+        public
+        view
+        returns (
+            uint64,
+            uint64,
+            uint256
+        )
+    {
+        if (identities[id].length == 0) {
+            return (0, 0, 0);
+        }
+        IDState memory lastIdState = identities[id][identities[id].length - 1];
+        return (
+            lastIdState.BlockN,
+            lastIdState.BlockTimestamp,
+            lastIdState.State
+        );
+    }
+
+    function getSmtCurrentRoot() public view returns (uint256) {
+        return smt.root();
```

Comment:
what happens is smt is not set? and what is expected behaviour?

Reply:
We haven't check it, actually. The transaction/call may be reverted, I think.